library/src/main/scala/scala/scalajs/js/Union.scala

```diff
@@ -0,0 +1,65 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js API               **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2015, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-lang.org/     **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package scala.scalajs.js
+
+import scala.language.implicitConversions
+
+/** Value of type A or B (union type).
+ *
+ *  Scala does not have union types, but they are important to many
+ *  interoperability scenarios. This type provides a (partial) encoding of
+ *  union types using implicit evidences.
+ */
+@scala.scalajs.js.annotation.RawJSType // Don't do this at home!
+sealed trait |[A, B] // scalastyle:ignore
+
+object | { // scalastyle:ignore
+  /** Evidence that `A <: B`, taking top-level `|`-types into account. */
+  sealed trait Evidence[-A, +B]
+
+  /** A unique (and typically dead-code-eliminated away) instance of
+   *  `Evidence`.
+   */
+  private object ReusableEvidence extends Evidence[scala.Any, scala.Any]
+
+  abstract sealed class EvidenceLowestPrioImplicits {
+    /** If `A <: B2`, then `A <: B1 | B2`. */
+    implicit def right[A, B1, B2](implicit ev: Evidence[A, B2]): Evidence[A, B1 | B2] =
+      ReusableEvidence.asInstanceOf[Evidence[A, B1 | B2]]
+  }
+
+  abstract sealed class EvidenceLowPrioImplicits extends EvidenceLowestPrioImplicits {
+    /** `Int <: Double`, because that's true in Scala.js. */
+    implicit def intDouble: Evidence[Int, Double] =
+      ReusableEvidence.asInstanceOf[Evidence[Int, Double]]
```

Comment:
I mean I call `|` a "union type", but it does have a lot of limitations compared to a *true* union type as you would find in dotty, obviously.

I would suggest we keep it minimal first. As people start using it, they'll complain that this or that use case is not supported, and we can add more implicits, until most/all common cases are supported. WDYT?